src/hardware/gameblaster.cpp

```diff
@@ -133,21 +133,24 @@ bool GameBlaster::MaybeRenderFrame(AudioFrame &frame)
 	static int16_t *p_buf[]           = {&buf[0], &buf[1]};
 	static device_sound_interface::sound_stream stream;
 
+	int16_t leftAccum = 0;
+	int16_t rightAccum = 0;
+
 	// Accumulate the samples from both SAA-1099 devices
 	devices[0]->sound_stream_update(stream, 0, p_buf, 1);
-	frame.left  = buf[0];
-	frame.right = buf[1];
+	leftAccum = buf[0];
+	rightAccum = buf[1];
 
 	devices[1]->sound_stream_update(stream, 0, p_buf, 1);
-	frame.left += buf[0];
-	frame.right += buf[1];
+	leftAccum += buf[0];
+	rightAccum += buf[1];
 
 	// Increment our time datum up to which the device has rendered
 	last_rendered_ms += ms_per_render;
 
 	// Resample the limited frame
-	const auto l_ready = resamplers[0]->input(buf[0]);
-	const auto r_ready = resamplers[1]->input(buf[1]);
+	const auto l_ready = resamplers[0]->input(leftAccum);
+	const auto r_ready = resamplers[1]->input(rightAccum);
```

Comment:
Very nice catch, @canadacow.

We previously soft-limited from float and wrote the result back into the 16-bit `buf`, preventing the dropping of the bottom channels, but this was removed in commit: https://github.com/dosbox-staging/dosbox-staging/commit/01fe221ba002a824472c2eb19ee4a57b1e05b7b3; thanks for catching this regression!

Reply:
Understandable since the CMS/Gameblaster is notoriously bad sounding anyway. I actually had to go online to listen to actually recordings to confirm my memory as to the sound of the Gameblaster. Sure enough, only half of the badness was there! :-)

Reply:
> Sure enough, only half of the badness was there! :-)

:joy:

Reply:
Thanks for pitching in your help - and I see you're working on other interesting goodies.  We're especially focused on the basic DOS years, with X and PCem/x86box doing the heavy for those win95+ years. Bring on that badness :laughing:

Reply:
You're very welcome! My fork of dosbox-staging is mainly to get HDR support in CRT shaders. https://github.com/canadacow/dosbox-staging/tree/canadacow-edits

So far I've upgraded SDL2 to support HDR swapchains (R16G16B16A16_FLOAT) and I modified the SDL2 texture output to use a DX11 device to run CRT shaders ported to compute. The end result is HDR emulation of the shadow mask in classic CRTs.

Thanks for your hard work brining modern practices to the DOSBox codebase. (Bit of trivia: The IMGMOUNT command, extremely early CMS/Gameblaster code and, IPX and the modem emulation are all my contributions to DOSBox.)

Reply:
> Thanks for your hard work bringing modern practices to the DOSBox codebase.

You're very welcome!

> Bit of trivia

:handshake: no doubt, DOSBox wouldn't be what it without your contributions!

Everyone's work is right here, as one continuum, starting from SVN #1 :+1: 

2dbe07074 18 years ago Dean Beeler Completed S3 XGA compatibility
d1d89dc80 18 years ago Dean Beeler Completing S3 XGA functionality.  Fixed minor mouse glitch.
1d46933a6 18 years ago Dean Beeler Completing S3 XGA functionality.
26062b0e1 18 years ago Dean Beeler Adding serial port FOSSIL driver support
b59b4b679 18 years ago Dean Beeler Fixing fat write access.  Still no DOS error codes yet, sorry.
4162c2b7a 18 years ago Dean Beeler Added IMGMOUNT, BOOT and IPX documentation
7a1ab977c 18 years ago Dean Beeler Adding Win32 passthrough serial support (late)
23005cba5 18 years ago Dean Beeler Adding Win32 passthrough serial support
f84108170 18 years ago Dean Beeler Added FAT image access and DOS booter functionality
d0e2bfa15 18 years ago Dean Beeler Added FAT image access and DOS booter functionality
57a7795ae 18 years ago Dean Beeler Fix default mixer sampling rate.
45e55eb8d 19 years ago Dean Beeler Added config support for building with IPX
ddba2a34f 19 years ago Dean Beeler Fixed modem to work with the Telnet protocol.  Integrated fizzban's code to dial a striing of numbers as an IP.  This release should work with Xmodem and Ymodem protocols.
7ee4be6fd 19 years ago Dean Beeler Added initial IPX support
8dda3577a 19 years ago Dean Beeler Added external SDLNetInited variable so the modem and IPX stuff don't cause problems when both enabled.
c27159214 19 years ago Dean Beeler Added config support for building with IPX
143bad813 19 years ago Dean Beeler Canadacow changes for 0.61
d38d9a014 19 years ago Dean Beeler Canadacow changes for 0.61
0bec45e6b 19 years ago Dean Beeler Fixed channel iterating
9f1e9e305 19 years ago Dean Beeler Change to ensure correct number of channels played
6a3c799e1 19 years ago Dean Beeler Tuned 16-bit samples.  Increased support for some tracker programs.
909f7333f 19 years ago Dean Beeler Restored lost DMA functionality.  (Fixed MM45 bug.)
247678bbe 19 years ago Dean Beeler Added documentation.  Removed outdated code.
4239e8d87 19 years ago Dean Beeler Moved DMA channel code to dma.cpp
ed4d000c1 19 years ago Dean Beeler Implemented 16-bit DMA support
9d8d93a7d 19 years ago Dean Beeler Added GUS init section
4057f5005 19 years ago Dean Beeler Added port hack to Gravis to aid in GUS detection
f9cb7aa8d 19 years ago Dean Beeler Added GUS functionality

Reply:
> My fork of dosbox-staging is mainly to get HDR support in CRT shaders.

I have to warn you that there are plans to remove surface output and scalers, rewrite window routines and output pipeline, and replace current shaders system with RetroArch's one. Just so you don't spend/waste more time than necessary.

Regarding IPX, I'd like to ask: Did you have any problems related to higher network latency in games? We're facing periodic stutters and slowdowns in tested games, and some games may even crash and burn hard (Rise of the Triad). Curious if you could provide some insight.

(Terribly sorry for hijacking this thread.)

Reply:
@GranMinigun , continuing the chat over discord...